src/tools/rbd/Utils.cc

```diff
@@ -839,17 +870,49 @@ int open_image(librados::IoCtx &io_ctx, const std::string &image_name,
   return 0;
 }
 
+int open_image_by_id(librados::IoCtx &io_ctx, const std::string &image_id,
+                     bool read_only, librbd::Image *image) {
+  int r;
+  librbd::RBD rbd;
+  if (read_only) {
+    r = rbd.open_by_id_read_only(io_ctx, *image, image_id.c_str(), NULL);
+  } else {
+    r = rbd.open_by_id(io_ctx, *image, image_id.c_str());
+  }
+
+  if (r < 0) {
+    std::cerr << "rbd: error opening image with id " << image_id << ": "
+              << cpp_strerror(r) << std::endl;
+    return r;
+  }
+  return 0;
+}
+
 int init_and_open_image(const std::string &pool_name,
                         const std::string &image_name,
                         const std::string &snap_name, bool read_only,
                         librados::Rados *rados, librados::IoCtx *io_ctx,
                         librbd::Image *image) {
+  return init_and_open_image_by_id(pool_name, image_name, "", snap_name,
+                                   read_only, rados, io_ctx, image);
+}
+
+int init_and_open_image_by_id(const std::string &pool_name,
```

Comment:
Nit: the ```_by_id``` suffix seems off to me since it can open by name as well. I would just drop it and always open by id as first choice if it's provided.